File: src/main/java/fr/raconteur/simpleskinswapper/gui/SkinPreviewButton.java
package fr.raconteur.simpleskinswapper.gui;

import fr.raconteur.simpleskinswapper.changeskin.SelectedSkinStore;
import net.minecraft.client.gui.DrawContext;
import net.minecraft.client.gui.widget.ButtonWidget;
import net.minecraft.entity.player.PlayerSkinType;
import net.minecraft.entity.player.SkinTextures;
import net.minecraft.util.AssetInfo;
import net.minecraft.util.Identifier;

public class SkinPreviewButton extends ButtonWidget {

    private final int previewWidth;
    private final int previewHeight;
    private final int previewGap;

    public SkinPreviewButton(int x, int y, int width, int height,
                              int previewWidth, int previewHeight, int previewGap,
                              net.minecraft.text.Text message, ButtonWidget.PressAction onPress) {
        super(x, y, width, height, message, onPress, DEFAULT_NARRATION_SUPPLIER);
        this.previewWidth = previewWidth;
        this.previewHeight = previewHeight;
        this.previewGap = previewGap;
    }

    @Override
    protected void drawIcon(DrawContext context, int mouseX, int mouseY, float delta) {
        this.drawButton(context);
        this.drawLabel(context.getHoverListener(this, DrawContext.HoverType.NONE));
        this.drawPreview(context, mouseX, mouseY);
    }

    private void drawPreview(DrawContext context, int mouseX, int mouseY) {
        Identifier texture = SelectedSkinStore.getPreviewTexture();
        if (texture == null) return;

        SkinType skinType = SelectedSkinStore.getPreviewSkinType();
        PlayerSkinType modelType = (skinType == SkinType.SLIM) ? PlayerSkinType.SLIM : PlayerSkinType.WIDE;
        SkinTextures skin = new SkinTextures(
                new AssetInfo.SkinAssetInfo(texture, ""), null, null,
                modelType, true
        );

        int centerX = this.getX() + this.getWidth() / 2;
        int y2 = this.getY() - previewGap;
        int y1 = y2 - previewHeight;
        int x1 = centerX - previewWidth / 2;
        int x2 = centerX + previewWidth / 2;

        SkinRenderer.renderPlayerFollowingMouse(context, x1, y1, x2, y2, previewHeight / 2, skin, mouseX, mouseY);
    }
}

File: src/main/kotlin/fr/raconteur/simpleskinswapper/gui/SkinPreviewButton.kt
package fr.raconteur.simpleskinswapper.gui

import fr.raconteur.simpleskinswapper.changeskin.SelectedSkinStore
import net.minecraft.client.gui.GuiGraphicsExtractor
import net.minecraft.client.gui.components.Button
import net.minecraft.core.ClientAsset
import net.minecraft.network.chat.Component
import net.minecraft.world.entity.player.PlayerModelType
import net.minecraft.world.entity.player.PlayerSkin

class SkinPreviewButton(
    x: Int, y: Int, width: Int, height: Int,
    private val previewWidth: Int,
    private val previewHeight: Int,
    private val previewGap: Int,
    message: Component,
    onPress: OnPress
) : Button(x, y, width, height, message, onPress, Button.DEFAULT_NARRATION) {

    override fun extractContents(graphics: GuiGraphicsExtractor, mouseX: Int, mouseY: Int, a: Float) {
        this.extractDefaultSprite(graphics)
        this.extractDefaultLabel(graphics.textRendererForWidget(this, GuiGraphicsExtractor.HoveredTextEffects.NONE))
        this.renderPreview(graphics, mouseX, mouseY)
    }

    private fun renderPreview(graphics: GuiGraphicsExtractor, mouseX: Int, mouseY: Int) {
        val texture = SelectedSkinStore.getPreviewTexture() ?: return

        val skinType = SelectedSkinStore.getPreviewSkinType()
        val modelType = if (skinType == SkinType.SLIM) PlayerModelType.SLIM else PlayerModelType.WIDE
        val skin = PlayerSkin(
            ClientAsset.DownloadedTexture(texture, ""), null, null,
            modelType, true
        )

        val centerX = this.x + this.width / 2
        val y2 = this.y - previewGap
        val y1 = y2 - previewHeight
        val x1 = centerX - previewWidth / 2
        val x2 = centerX + previewWidth / 2

        SkinRenderer.renderPlayerFollowingMouse(graphics, x1, y1, x2, y2, previewHeight / 2, skin, mouseX, mouseY)
    }
}
